pkg/start/start.go

```diff
@@ -153,16 +178,97 @@ func (o *Options) Run() error {
 	return nil
 }
 
+func (o *Options) handler() http.Handler {
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.Handler())
+	return mux
+}
+
+func (o *Options) makeTLSConfig() (*tls.Config, error) {
+	if !fileExists(o.ServingCertFile) {
+		return nil, errors.New(o.ServingCertFile + " does not exist")
+	}
+
+	if !fileExists(o.ServingKeyFile) {
+		return nil, errors.New(o.ServingKeyFile + " does not exist")
+	}
+
+	// Load the initial certificate contents.
+	certBytes, err := ioutil.ReadFile(o.ServingCertFile)
+	if err != nil {
+		return nil, err
+	}
+	keyBytes, err := ioutil.ReadFile(o.ServingKeyFile)
+	if err != nil {
+		return nil, err
+	}
+	certificate, err := tls.X509KeyPair(certBytes, keyBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return crypto.SecureTLSConfig(&tls.Config{
+		GetCertificate: func(_ *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			return &certificate, nil
+		},
+		ClientAuth: tls.RequestClientCert,
+	}), nil
+}
+
 func (o *Options) run(ctx context.Context, controllerCtx *Context, lock *resourcelock.ConfigMapLock) {
 	// listen on metrics
 	if len(o.ListenAddr) > 0 {
-		mux := http.NewServeMux()
-		mux.Handle("/metrics", promhttp.Handler())
+		handler := o.handler()
+		tcpl, err := net.Listen("tcp", o.ListenAddr)
+		if err != nil {
+			klog.Errorf("Listen error: %v", err)
+		}
+
+		// if a TLS connection was requested, set up a connection mux that will send TLS requests to
+		// the TLS server but send HTTP requests to the HTTP server. Preserves the ability for HTTP
+		// health checks to call HTTP on the router while still allowing TLS certs and end to end
+		// metrics protection.
+		m := cmux.New(tcpl)
+
+		// match HTTP first
+		httpl := m.Match(cmux.HTTP1())
 		go func() {
-			if err := http.ListenAndServe(o.ListenAddr, mux); err != nil {
-				klog.Fatalf("Unable to start metrics server: %v", err)
+			s := &http.Server{
+				Handler: handler,
+			}
+			if err := s.Serve(httpl); err != cmux.ErrListenerClosed {
+				klog.Errorf("Serve error (http): %v", err)
 			}
 		}()
+
+		var TLSConfig *tls.Config
+		if tlsConfig, err := o.makeTLSConfig(); err != nil {
+			klog.Errorf("Could not create tls Config: %v", err)
```

Comment:
nit: `tls Config` -> `TLS config` and `Errorf` -> `Fatalf`.  Maybe `Could not` -> `Failed to`, because I sometimes grep for "fail" in logs but rarely grep for "could not" ;).

Reply:
@wking Changing to Fatalf causes failures. I had hoped that it would go away with https://github.com/openshift/cluster-version-operator/pull/366 but apparently not. It seemed similar but what I originally believe was happening was the pod was coming up before the mount was there causes it to be restarted afterwhich the file is there. With the Fatalf it just get killed off and not restarted. That was the best theory I had at the time in any case.
So when it has Fatalf the fails are at [1405](https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/358/pull-ci-openshift-cluster-version-operator-master-e2e-aws/1405) and [772](https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/358/pull-ci-openshift-cluster-version-operator-master-e2e-aws-upgrade/772) . Unfortunately the must-gathers fail as well.

Reply:
fixed in abf62a5